### PR TITLE
Replace use of fstat()

### PR DESCRIPTION
**Pull request by ccawley2011**

### CODE REVIEW

File: src/depackers/xfd.c [hunk 1]
```diff
@@ -57,8 +56,7 @@ static int decrunch_xfd(FILE *f1, FILE *f2)
     if (f2 == NULL)
 	return -1;
 
-    fstat(fileno(f1), &st);
```
Comment (sezero) [inline src/depackers/xfd.c:60]: Lose the local var `struct stat st;` too (at line 51)

**Comment by sezero** [comment]
I actually like this one. (If only we could get rid of stat() use load.c, too..)

@AliceLR what do you think.

**Review by sezero — APPROVED** [review]

**Comment by ccawley2011** [comment]
> (If only we could get rid of stat() use `load.c` and `loaders/mfp_load.c`, too..)

I got rid of the ones in `loaders/mfp_load.c`, but the tests currently rely on `xmp_load_module` and `xmp_test_module` explicitly setting `errno` to `EISDIR`.

**Comment by sezero** [comment]
I have os-specific small procs (windows, os2, djgpp, amiga, and unix
which does need stat) that return file, dir or none for a given path, so
setting errno to EISDIR shouldn't be a problem.

**Comment by AliceLR** [comment]
I think using the `SEEK_END` hack instead of `fstat` is a terrible idea—it is slower, clobbers the file buffer, and [library implementations are allowed to not meaningfully support `SEEK_END` for binary streams](https://www.cplusplus.com/reference/cstdio/fseek/), thus it is inherently non-portable.

**Review by AliceLR — CHANGES_REQUESTED** [review]
See above.

**Comment by sezero** [comment]
You're actually right.

**Comment by sezero** [comment]
This actually seems like a can of worms. I don't like `fstat` in anywhere
other than unix, but querying an open `FILE*` without it can be a pain
otherwise.

**Comment by Clownacy** [comment]
Personally, I've never understood the issue of `SEEK_END` not being guaranteed to actually do anything: it sounds to me like a goofy excuse for the file IO API to support networks streams and `stdin` and other things which physically lack an end. In practice, is a platform _really_ not going to implement `SEEK_END` for a traditional filesystem?

**Comment by AliceLR** [comment]
I haven't encountered anything that doesn't actually support `SEEK_END` yet.

I'm not sure this is actually that bad: here's roughly the handling for this I've used elsewhere, and I've yet to encounter issues. `_filelength` takes precedence mainly because some ad hoc testing I did suggested it's unsurprisingly faster than `fstat` in Windows.

```C
struct stat st;
int fd = fileno(f);
long size;
long pos;

#ifdef _WIN32

size = _filelength(fd);
if(size >= 0)
  return size;

#endif

if(!fstat(fd, &st))
  return st.st_size;

pos = ftell(f);
if(pos < 0)
  return pos;

if(fseek(f, 0, SEEK_END))
  return -1;

size = ftell(vf);
if(fseek(vf, pos, SEEK_SET))
  return -1;

return size;
```

**Comment by sezero** [comment]
`_filelength` seems to call `lseek` with `SEEK_END`

**Comment by sezero** [comment]
This patchset actually removes `fstat` _mainly_ from depackers, yes?
And we always feed depack() the `FILE*` from hio, yes? We already have
the file length cached in hio, retrievable with `hio_size`: How about we
change `struct depacker->depack()` to have a third `long flen` parameter
sent as `hio_size()`?

That way, the only `fstat` users left will be `test_depack_vorbis.c` and 
`test_depack_vorbis_8bit.c`, well apart from `hio.c` itself which can be
managed separately?

**Comment by AliceLR** [comment]
> This patchset actually removes `fstat` _mainly_ from depackers, yes?
> And we always feed depack() the `FILE*` from hio, yes? We already have
> the file length cached in hio, retrievable with `hio_size`: How about we
> change `struct depacker->depack()` to have a third `long flen` parameter
> sent as `hio_size()`?

I think that's reasonable—it looks like adding that to all affected depackers would be non-intrusive (and might allow further non-intrusive checks in other depackers).

**Comment by sezero** [comment]
OK, here is an initial patch that adds an `inlen` to depack() and removes
`fstat` use from depackers: [patch_442_1.txt](https://github.com/libxmp/libxmp/files/6854762/patch_442_1.txt)

`test_depack_vorbis.c` and `test_depack_vorbis_8bit.c` of test-dev still use
fstat: to be handled.

I think Cameron's patch to `mfp_load.c` in https://github.com/libxmp/libxmp/pull/442/commits/4d28416e3f08732ca2fb984c341f3a46b32ab3e9 is OK: what do you think?

**Comment by AliceLR** [comment]
> OK, here is an initial patch that adds an `inlen` to depack() and removes
> `fstat` use from depackers: [patch_442_1.txt](https://github.com/libxmp/libxmp/files/6854762/patch_442_1.txt)

This patch looks OK to me.

> `test_depack_vorbis.c` and `test_depack_vorbis_8bit.c` of test-dev still use
> fstat: to be handled.

FWIW I think the general approach in Cameron's patch (creating and exposing a function for these) is OK, I was more concerned about the methodology of only relying on that particular hack to get the file size. `_filelength` using the seek hack internally makes sense since I remember those two methods having similar performance.

> I think Cameron's patch to `mfp_load.c` in [4d28416](https://github.com/libxmp/libxmp/commit/4d28416e3f08732ca2fb984c341f3a46b32ab3e9) is OK: what do you think?

I think the MFP loader changes are OK. (Probably out of scope, but the sample name should probably be sanitized through `libxmp_copy_name_for_fopen`.)

**Comment by sezero** [comment]
And here is a patch to test-dev side that changes two depack_vorbis
tests to use `read_file_to_memory`, and `test_write_file_move_data` to
use `SEEK_END` and `ftell` after read is done for sake of simplicity.
[patch_442_2.txt](https://github.com/libxmp/libxmp/files/6855482/patch_442_2.txt)

**Comment by sezero** [comment]
OK then,
- my patch-1 is OK to apply
- Cameron's mfp patch is OK to apply, then you'll push revisions
- is my patch-2 OK?

**Comment by AliceLR** [comment]
> * my patch-1 is OK to apply
> * Cameron's mfp patch is OK to apply, then you'll push revisions
> * is my patch-2 OK?

That is OK with me.

**Comment by sezero** [comment]
Patches applied as https://github.com/libxmp/libxmp/commit/f5d8242c0782305dc97a0076c6c1a1a9f4590b8b, https://github.com/libxmp/libxmp/commit/a9fd33a21f2c2b61c2c61646a0efe97e968b5439, and https://github.com/libxmp/libxmp/commit/3f69e19da41548fa171036dca4bce9e8f3d62d4e.

**Pull request closed without merging**